btree/btree.c

```diff
@@ -3529,21 +3566,417 @@ int64_t btree_iter_tick(struct m0_sm_op *smop)
-			return P_DONE;
+	if (cookie_is_used() || oi->i_used == 0)
+		return true;
 
-		default:
-			return 0;
+	l_node = oi->i_level[1].l_sib;
+
+	if (l_node) {
+		if (!node_is_valid(l_node))
+			return false;
+		if (oi->i_level[1].l_sib_seq != l_node->n_seq)
+			return false;
+	}
+	return true;
+}
+
+/**
+ * This function will determine if there is requirement of loading root child.
+ * If root contains only two records and if any of them is going to get deleted,
+ * it is required to load the other child of root as well to handle root case.
+ *
+ * @param bop it will provide all required information about btree operation.
+ * @return int8_t return -1 if any ancestor node is not valid. return 1, if
+ *                loading of child is needed, else return 0;
+ */
+static int8_t root_child_is_req(struct m0_btree_op *bop)
+{
+	struct m0_btree_oimpl *oi = bop->bo_i;
+	int8_t load = 0;
+	int used_count = oi->i_used;
+	do {
+		if (!node_is_valid(oi->i_level[used_count].l_node))
+			return -1;
+		if (used_count == 0) {
+			load = 1;
+			break;
+		}
+		if (!node_underflow(oi->i_level[used_count].l_node, true))
+			break;
+
+		used_count--;
+	}while (1);
+	return load;
+}
+
+/**
+ * This function will get called if root is an internal node and it contains
+ * only two records. It will check if there is requirement for loading root's
+ * other child and accordingly return the next state for execution.
+ *
+ * @param bop it will provide all required information about btree operation.
+ * @return int64_t it return state which needs to get executed next.
+ */
+static int64_t handle_root_case(struct m0_btree_op *bop)
+{
+	/**
+	 * If root is an internal node and it contains only two records, check
+	 * if any record is going to be deleted if yes, we also have to load
+	 * other child of root so that we can copy the content from that child
+	 * at root and decrease the level by one.
+	 */
+	struct m0_btree_oimpl *oi = bop->bo_i;
+	int8_t                 load;
+
+	load = root_child_is_req(bop);
+	if (load == -1) {
+		level_cleanup(oi, bop->bo_tx);
+		return P_SETUP;
+	}
+	if (load) {
+		struct slot    root_slot = {};
+		struct segaddr root_child;
+		struct level   *root_lev = &oi->i_level[0];
+
+		root_slot.s_node = root_lev->l_node;
+		root_slot.s_idx  = root_lev->l_idx == 0 ? 1 : 0;
+
+		node_child(&root_slot, &root_child);
+		if (!address_in_segment(root_child)) {
+			node_op_fini(&oi->i_nop);
+			return fail(bop, M0_ERR(-EFAULT));
+		}
+
+		return node_get(&oi->i_nop, bop->bo_arbor->t_desc,
+				&root_child, P_STORE_CHILD);
+	}
+	return P_LOCK;
+}
+
+/* State machine implementation for delete operation */
+static int64_t btree_del_tick(struct m0_sm_op *smop)
+{
+	struct m0_btree_op    *bop        = M0_AMB(bop, smop, bo_op);
+	struct td             *tree       = bop->bo_arbor->t_desc;
+	uint64_t               flags      = bop->bo_flags;
+	struct m0_btree_oimpl *oi         = bop->bo_i;
+	struct level          *lev;
+
+	if (oi)
+		lev = &oi->i_level[oi->i_used];
```

Comment:
It looks weird at this place. Please initialise it at the specific states wherever required.

Reply:
ok